resources/testharness.js

```diff
@@ -532,30 +531,34 @@ policies and contribution forms [3].
 
     function promise_test(func, name, properties) {
         var test = async_test(name, properties);
+        test._is_promise_test = true;
+
         // If there is no promise tests queue make one.
         if (!tests.promise_tests) {
             tests.promise_tests = Promise.resolve();
         }
         tests.promise_tests = tests.promise_tests.then(function() {
             var donePromise = new Promise(function(resolve) {
-                test._add_cleanup(resolve);
+                test._add_done_callback(resolve);
             });
             var promise = test.step(func, test, test);
+
             test.step(function() {
                 assert_not_equals(promise, undefined);
             });
-            Promise.resolve(promise).then(
-                    function() {
-                        test.done();
-                    })
+
+            Promise.resolve(promise)
```

Comment:
Exactly

Reply:
Practically speaking, yes. Technically, the semantics are slightly different,
though.

- In `master` today, if the invocation of `test.done` itself produces an
  exception, then this will be reported as a test failure.
- With this patch applied, if the invocation of `test.done` itself produces an
  exception, then this will cause a global "unhandled rejection" event, which
  will subsequently be reported as a harness failure.

Although this is not a condition we expect to ever occur, it seems better to
report the error as a harness failure rather than a test failure.

Reply:
That makes sense, thanks!

Reply:
Although perhaps it has something to do with the new harness errors that are seen with this change. Let's see about that...